### DISCUSSION TITLE
Memory issue when training in DDP mode

### DISCUSSION THREAD

**Discussion by happysyp000** [Lightning-AI/pytorch-lightning#18525] [DDP / multi-GPU / multi-node]
Hi, I have a training dataset about 80GB that is saved in a pandas dataframe. My machine has 4 GPUs, and the CPU memory is around 300Gb. When training the model with 1GPU, everything works fine, but when I train with 4 GPUs with the "ddp" mode, the memory gets dangerously close to overflown (for example, the memory usage would be around 320GB rather than 80GB)

My datamodule is like following:
```
class datamodule(pl.LightningDataModule):
    def __init__(dataframe_location):
      self.data_path=dataframe_location

    def setup(self, stage):
       if stage=='fit':
          self.df = pd.read_pickle(self.data_path)
  
    def train_dataloader(self):
          dataloader = loader_func(df,...)
          return dataloader
```

My impression is the ddp mode creates 4 copies of the datamodule when training on GPUs, thus loading the dataframe 4 times, causing the memory issue. I also tried to put loading of the dataframe to the __init__ function of the datamodule, but the memory issue still occurs.

Since the 4 GPUs are trained with the same dataframe, I'd like to find a way so that the dataframe needs to be loaded only one time, and  keep the CPU from memory overflown. Is there any suggestion as to what I can do?

Thanks.

**Comment by My-captain** [comment]
same issue

**Comment by kiucho** [comment]
Same issue too.

When training facebook/opt-1.3b (BF16-mixed) with Lightning, there is a difference in vram usage depending on the strategy.
with strategy: "auto" it allocates 29GB which seems proper,
![image](https://github.com/Lightning-AI/pytorch-lightning/assets/91112750/27739756-67e3-43fe-863d-5610e0787655)

But with strategy: "ddp" it allocates 41GB per GPU.
![image](https://github.com/Lightning-AI/pytorch-lightning/assets/91112750/1612e9d0-e906-458f-87b2-aa5290faaadf)

Does anyone know what might be causing this?

**Comment by KasuganoLove** [comment]
Here is a similar question and with an answer. I think it is helpful.

> https://github.com/Lightning-AI/pytorch-lightning/discussions/11763
